core/src/main/scala/org/apache/spark/sql/execution/joins/LocalJoin.scala

```diff
@@ -112,22 +116,41 @@ case class LocalJoin(leftKeys: Seq[Expression],
 
     // materialize dependencies in the entire buildRDD graph for
     // buildRDD.iterator to work in the compute of mapPartitionsPreserve below
-    materializeDependencies(buildRDD, new mutable.HashSet[RDD[_]]())
-
-    val schema = buildPlan.schema
-    streamRDD.mapPartitionsPreserveWithPartition { (context, _, itr) =>
-      val start = System.nanoTime()
-      val hashed = HashedRelationCache.get(schema, buildKeys, buildRDD,
-        buildPartition, context)
-      buildTime += (System.nanoTime() - start) / 1000000L
-      val estimatedSize = hashed.estimatedSize
-      buildDataSize += estimatedSize
-      context.taskMetrics().incPeakExecutionMemory(estimatedSize)
-      context.addTaskCompletionListener(_ => hashed.close())
-      join(itr, hashed, numOutputRows)
+    if (buildRDD.partitions.size == 1) {
+      materializeDependencies(buildRDD, new mutable.HashSet[RDD[_]]())
+      val schema = buildPlan.schema
+      streamRDD.mapPartitionsPreserveWithPartition { (context, split, itr) =>
+        val start = System.nanoTime()
+        val hashed = HashedRelationCache.get(schema, buildKeys, buildRDD,
+          buildRDD.partitions(0), context)
+        buildTime += (System.nanoTime() - start) / 1000000L
+        val estimatedSize = hashed.estimatedSize
+        buildDataSize += estimatedSize
+        context.taskMetrics().incPeakExecutionMemory(estimatedSize)
+        context.addTaskCompletionListener(_ => hashed.close())
+        join(itr, hashed, numOutputRows)
+      }
+    } else {
+      streamRDD.zipPartitions(buildRDD) { (streamIter, buildIter) =>
+        val hashed = buildHashedRelation(buildIter)
+        join(streamIter, hashed, numOutputRows)
+      }
```

Comment:
Actually in doExecute() I think we should just do a WholeStageCodegen(this).execute() since there is no case in which code generation of this plan should fail (apart from a bug that is). I am not 100% sure about this but we can check that after release (e.g. see BatchedDataSourceScanExec.doExecute). It makes things simpler for us with less surprises -- we are already quite a surprised bunch :-)